---
title: "Bayesian analysis to associate within-population genetic variation with potential underlying drivers"
subtitle: "Based on real data"
author: "Juliette Archambeau"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    # code_fold: hide
    toc: true
    toc_depth: 4
    toc_float:
       collapsed: false
    number_sections: true
    highlight: textmate
bibliography: references.bibtex
editor_options: 
  chunk_output_type: console
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 600px;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5,fig.height = 4,cache=F)
options(width = 300)
library(knitr)        # CRAN v1.26
library(reshape2)     # CRAN v1.4.4
library(dplyr)        # CRAN v1.0.0
library(tidyverse)    # CRAN v1.3.0
library(kableExtra)   # CRAN v1.1.0
library(rstan)        # CRAN v2.19.2
library(broom)        # CRAN v0.5.2
library(loo)          # CRAN v2.2.0
library(latex2exp)    # CRAN v0.4.0
library(parallel)
library(tidytidbits)  # CRAN v0.2.3
library(tidybayes)    # CRAN v2.0.1
library(bayesplot)    # CRAN v1.7.1
library(cowplot)      # CRAN v1.0.0
library(xtable)
theme_set(theme_bw(base_size = 20))
color_scheme_set("green")
rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())
```


```{r OptionDocs}
# Options for sampling
n_chains <- 4
n_iter <- 2500
n_warm <- 1250
n_thin <- 1
save_warmup = FALSE 

theme_set(theme_bw(base_size = 20))

# option for the MCMC intervals
point.est="median"

conf.level <- 0.95
```

# Model equation and code

In this document, we aimed to estimate the relationship between potential drivers (namely, two admixture scores, four indexes of environmental heterogeneity and two indexes of climatic harshness) and the within-population genetic variation. 

We modeled each trait $y_{bpcr}$ such as:

\begin{equation}
\begin{aligned}
y_{bpcr} & \sim  \mathcal{N}(\mu_{bpc},\sigma^{2}_{r})\\[3pt]
\mu_{bpc} & =  \beta_{0} +  B_{b} + P_{p} + C_{c(p)}\\[3pt]
\end{aligned}
\end{equation}

where $\beta_{0}$ is the global intercept, $B_{b}$ the block intercepts, $P_{p}$ the population intercepts, $C_{c(p)}$ the clone intercepts and $\sigma^{2}_{r}$ the residual variance.

The prior of $\beta_{0}$ was weakly informative and centered around the mean of the observed values for the trait under considered, as follows:

$$\beta_{0} \sim \mathcal{N}(\mu_{y},2)$$

The population and block intercepts, $P_{p}$ and $B_{b}$ were considered normally-distributed with variances $\sigma^{2}_{P}$ and $\sigma^{2}_{B}$, such as:

$$\begin{bmatrix} B_{b} \\ P_{p} \end{bmatrix} 
    \sim \mathcal{N}\left(0,
\begin{bmatrix}\sigma^{2}_{B}\\[3pt]
\sigma^{2}_{P}\\
\end{bmatrix}
\right)\\[3pt]$$

The clone intercepts $C_{c(p)}$ were considered to follow some population-specific normal distributions, such as:
$$C_{c(p)} \sim \mathcal{N}(0,\sigma^{2}_{C_{p}})$$

where $\sigma^{2}_{C_{p}}$ are the population-specific variances among clones.

To partition the total variance, we parameterize our model so that only the total variance, $\sigma_{tot}^{2}$ has a prior, such that:

\begin{equation}
\begin{aligned}
\sigma_{tot}^{2} & = \sigma_{r}^{2} + \sigma_{B}^{2} + \overline{\sigma_{C_{p}}^{2}} + \sigma_{P}^{2}\\[3pt]
\sigma_{r} & = \sigma_{tot} \times \sqrt(\pi_{r})\\[3pt]
\sigma_{B} & = \sigma_{tot} \times \sqrt(\pi_{B})\\[3pt]
\sigma_{P} & = \sigma_{tot} \times \sqrt(\pi_{P})\\[3pt]
\overline{\sigma_{C_{p}}} & = \sigma_{tot} \times \sqrt(\pi_{C})\\[3pt]
\sigma_{tot} & \sim \mathcal{S}^{*}(0,1,3)
\end{aligned}
\end{equation}

where $\overline{\sigma_{C_{p}}}$ and $\overline{\sigma_{C_{p}}^{2}}$ are the mean of the population-specific among-clones standard deviations ($\sigma_{C_{p}}$) and variances ($\sigma^{2}_{C_{p}}$), respectively, and $\sum_{l}^{4}\pi_{l}=1$ (using the `simplex` function in `Stan`).

The population-specific among-clones standard deviations $\sigma_{C_{p}}$ follow a log-normal distribution with mean $\overline{\sigma_{C_{p}}}$ and variance $\sigma^{2}_{K}$, such as: 

\begin{equation*}
\begin{aligned}
\sigma_{C_{p}} & \sim \mathcal{LN}\left(\ln(\overline{\sigma_{C_{p}}})-\frac{\sigma^{2}_{K}}{2} + \beta_{X}X_{p},\sigma^{2}_{K}\right)\\[3pt]
\sigma_{K} & \sim \exp(1)\\[3pt]
\end{aligned}
\end{equation*}

with $X_{p}$ the potential driver considered and $\beta_{x}$ its associated coefficient.


Here is the stan model code:

```{r ModelCompilation}
model = stan_model("scripts/StanModels/HierarchicalModel_VaryingInterClonesSD_OneCovariate.stan")
model
```

`r knitr::opts_chunk$set(eval = F)`

# Functions used

```{r FunctionDownloadFormatData,eval=F}
DownloadData <- function(trait,sub){

#### Trait transformations
    # Phenology-realted traits and delta13C are centered to help convergence
if(trait=="MeanBB"|trait=="MeanDBB"|trait=="d13C"){
    CenterResponseVariable <- TRUE
    LogResponseVariable <- FALSE
    
    # SLA is not centered and we used its logarithm as response variable
  }  else if(trait=="POR_SLA"){ 
    CenterResponseVariable <- FALSE
    LogResponseVariable <- TRUE
    
    # Height is not transformed
  } else { 
    CenterResponseVariable <- FALSE
    LogResponseVariable <- FALSE
  }

  
#### Downloading and scaling (mean-centering) the potential drivers
  drivers <- readRDS(file="data/DF_Drivers.rds") %>% 
    dplyr::select(prov,A,D,EH.20km.PC1,EH.20km.PC2,EH.1km.PC1,EH.1km.PC2,SHM,invEMT) %>% 
    dplyr::mutate(across(-prov,scale))

  
  
#### Merging with the phenotypic data
  if(trait=="MeanBB"|trait=="MeanDBB"){
      
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    when(trait=="MeanBB" ~ dplyr::select(.,prov,clon,tree,block,BDX_s32013,BDX_s32014,BDX_s32015,BDX_s32017),
         ~ dplyr::select(.,prov,clon,tree,block,BDX_dbb2014,BDX_dbb2015,BDX_dbb2017)) %>% 
    filter(rowSums(is.na(across(where(is.numeric)))) != ncol(.)-4) %>% 
    dplyr::mutate(trait = rowMeans(select(.,contains("BDX")),na.rm=T)) %>%
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov") 
  
  } else {
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    dplyr::rename(trait=all_of(trait)) %>% 
    dplyr::select(prov,clon,tree,block,trait) %>% 
    drop_na() %>%  
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov")
  
  }
  
  
########################################################################################################################
# Option to keep only populations represented by at least 4 clones & clones represented by at least 4 individuals.
  if(sub=="sub1"|sub=="sub2"){

  ClonesToKeep <- df %>% 
    dplyr::select(prov,clon) %>% 
    group_by(clon) %>% 
    mutate(NbInd = n()) %>% 
    distinct() %>%
    filter(NbInd>sub.ind)
  
  df <- df %>% semi_join(ClonesToKeep, by="clon") 
  
  PopsToKeep <- df %>%  
    dplyr::select(prov,clon) %>% 
    distinct() %>%  
    group_by(prov) %>%  
    dplyr::summarise(NbClones=n()) %>% 
    filter(NbClones>sub.clones)
  
  df <- df %>% semi_join(PopsToKeep, by="prov") 
  }
########################################################################################################################
    
  list.trait <- list(df=df)

#### Create the stan lists:
  list.stan <- lapply(colnames(drivers %>% dplyr::select(-prov)), function(x)    {
    list(N=length(df$tree),
         y=df$trait,
         X=as.numeric(pull(unique(df[,c(x,"prov")]) %>% dplyr::select(contains(x)))),
         nprov = length(unique(df$prov)),
         nclon = length(unique(df$clon)),
         nblock = length(unique(df$block)),
         prov = as.numeric(as.factor(df$prov)),
         which_prov = as.numeric(as.factor(pull(unique(df[c("prov","clon")])[,"prov"]))),
         clon = as.numeric(as.factor(df$clon)),
         bloc = as.numeric(as.factor(df$block)))})
  
  names(list.stan) <- colnames(drivers %>% dplyr::select(-prov))
  
  return(c(list.trait,list.stan))
}



ExtractParam <- function(x)  {
  ModelOutputs <- if(sub=="sub1"){
    readRDS(file=paste0("outputs/models/",x,"_sub1.rds"))} else if (sub=="sub2") {
    readRDS(file=paste0("outputs/models/",x,"_sub2.rds"))} else {
    readRDS(file=paste0("outputs/models/",x,".rds"))} 
  
  ModelOutputs %>% 
  mclapply(function(x) broom::tidyMCMC(x,pars=param,
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, 
                rhat = F, 
                conf.int = T,
                conf.level = conf.level)) %>% 
  bind_rows() %>% 
  mutate(Xp=drivers,
         Traits=x)
}
```

# Running and saving the models

For each of the eight trait, we ran eight models, one for each potential drivers:

```{r RunModel,eval=F}
########################################
# Running the models on a filtered dataset or not? (modification after the first round of review in Heredity)
sub = "no" # if the dataset is not filtered.

### Two filtered datasets:

# Subset 1: only populations represented by at least 4 clones and clones represented by at least 4 individuals.
sub = "sub1"
sub.clones = 3
sub.ind = 3

# Subset 1: only populations represented by at least 4 clones and clones represented by at least 4 individuals.
sub = "sub2"
sub.clones = 10
sub.ind = 6
########################################

if(sub=="sub2"){traits <- c("BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB")} else {
  traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")}



for(trait in traits){
list.trait <- DownloadData(trait,sub)

list.models <- list()

for(i in names(list.trait)[-1]){
  
  list.models[[i]] <- sampling(model, 
                               data = list.trait[[i]], 
                               pars=c("beta0","betaX",
                                      "pi","sigma2_r","sigma2_block","sigma2_prov","sigma2_clon",
                                      "h2_prov",
                                      "sigma_tot","sigma_block","sigma_prov",
                                      "mean_sigma_clon","sigma_clon","sigma_K",
                                      #"alpha_clon","alpha_prov",#"alpha_block",
                                      # "y_rep" not included not to have a too heavy file
                                      "R_squared","log_lik"),
                               iter = n_iter, 
                               chains = n_chains, 
                               cores = n_chains,
                               save_warmup = save_warmup,
                               thin=n_thin)
}

if(sub=="sub1"){
saveRDS(list.models,file=paste0("outputs/models/",trait,"_sub1.rds"))} else if (sub=="sub2"){
saveRDS(list.models,file=paste0("outputs/models/",trait,"_sub2.rds"))} else {
saveRDS(list.models,file=paste0("outputs/models/",trait,".rds"))}


}
```

We use the `loo` package to compare the models based on the [ELPD](http://mc-stan.org/loo/reference/loo-glossary.html).

```{r Loo,eval=F}
if(sub=="no"){ # we only ran this analysis for the unfiltered dataset
loos <- mclapply(list.models,loo)
loocompare <- loo_compare(x=loos)
saveRDS(loocompare,file=paste0("outputs/loo/",trait,".rds"))}
```

# Posterior distributions

## $\beta_{X}$

### One graph per trait

```{r BetaXTraitSpecificFigure,eval=F}
if(sub=="no"){
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT")# order in the model list, do not change
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")


mclapply(traits, function(x){
d <-readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) broom::tidyMCMC(x,pars=("betaX"),
              droppars = NULL, 
              estimate.method = point.est, 
              ess = F, 
              rhat = F, 
              conf.int = T,
              conf.level = conf.level)) %>% 
  bind_rows() %>% 
  mutate(Xp=drivers)
  
d$Xp <- as.factor(d$Xp)
d$Xp <- factor(d$Xp,levels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM"))


p <- ggplot(d,aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Xp)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=5,size=5) +
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_x_discrete(labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +

  scale_color_manual(values=c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" )) +
  #coord_flip() +
  theme_bw() +
  theme(axis.text = element_text(size=16),
        axis.title = element_text(size=16),
        legend.position = "none")


ggsave(p,file=paste0("figs/PosteriorDistributions/",x,"_BetaX.png"),height=8,width=12)

})}
```

### Main figure of the paper (Figure 2) (5 drivers)


```{r}
if(sub=="sub2"){traits <- c("BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB")} else {
  traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")}

drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT")   
subdrivers <- c("A","EH1[20km]","EH2[20km]","SHM","invEMT")
param <- "betaX"

df.betaX <- mclapply(traits,ExtractParam)

df.betaX <- df.betaX %>% 
  bind_rows() %>%
  dplyr::filter(Xp %in% subdrivers) %>% 
  mutate(Hyp=case_when(Xp=="A"~"Admixture",
                       Xp=="D"~"Admixture",
                       Xp=="EH1[20km]"~"Environmental heterogeneity",
                       Xp=="EH2[20km]"~"Environmental heterogeneity",
                       Xp=="EH1[1.6km]"~"Environmental heterogeneity",
                       Xp=="EH2[1.6km]"~"Environmental heterogeneity",
                       Xp=="SHM"~"Climate harshness",
                       Xp=="invEMT"~"Climate harshness"),
         TraitType=case_when(Traits=="POR_htoct12"~"Morphological",
                             Traits=="BDX_htnov13"~"Morphological",
                             Traits=="AST_htnov12"~"Morphological",
                             Traits=="BDX_htnov18"~"Morphological",
                             Traits=="MeanBB"~"Phenological",
                             Traits=="MeanDBB"~"Phenological",
                             Traits=="POR_SLA"~"Functional",
                             Traits=="d13C"~"Functional"))


df.betaX$Traits <- as.factor(df.betaX$Traits)
levels(df.betaX$Traits)
df.betaX$Traits <- factor(df.betaX$Traits,levels=rev(c(traits)))
df.betaX$Xp <- as.factor(df.betaX$Xp)
df.betaX$Xp <- factor(df.betaX$Xp,levels=rev(levels(df.betaX$Xp)))

d13c <- parse(text = TeX("$\\delta^{13}C$"))

if(sub=="sub2"){
  p <- df.betaX %>%   ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,show.legend = c(size = TRUE)) +
  facet_grid(.~Hyp,scales="free", space = "free") + 
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst")), # 
                     values=rev(c("#B2182B","#D73027", "#FDAE61","#5AAE61","#A6DBA0"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst")),
                     values = c(17, 17,16,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.342, .13),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=2))
} else {
p <- df.betaX %>%   ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,show.legend = c(size = TRUE)) +
  facet_grid(.~Hyp,scales="free", space = "free") + 
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)), # 
                     values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst")),
                     values = c(15, 15,17, 17,16,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.342, .13),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=2))
}

if(sub=="sub1"){
  ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_BetaX_PaperFig_sub1.png",height=8,width=10)} else if (sub=="sub2"){
  ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_BetaX_PaperFig_sub2.png",height=8,width=10)} else {
  ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_BetaX_PaperFig.png",height=8,width=10)  
  }
```


### Figure in Supp Info (5 drivers, all traits)


```{r AllTraits}
if(sub=="sub2"){traits <- c("BDX_htnov18","BDX_htnov13","AST_htnov12","MeanBB","MeanDBB")} else {
  traits <- c("BDX_htnov18","BDX_htnov13","POR_htoct12","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")}


drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT")   
param <- "betaX"

df.betaX <- mclapply(traits,ExtractParam)

df.betaX <- df.betaX %>% 
  bind_rows() %>%
  mutate(Hyp=case_when(Xp=="A"~"Admixture",
                       Xp=="D"~"Admixture",
                       Xp=="EH1[20km]"~"Environmental heterogeneity",
                       Xp=="EH2[20km]"~"Environmental heterogeneity",
                       Xp=="EH1[1.6km]"~"Environmental heterogeneity",
                       Xp=="EH2[1.6km]"~"Environmental heterogeneity",
                       Xp=="SHM"~"Climate harshness",
                       Xp=="invEMT"~"Climate harshness"),
         TraitType=case_when(Traits=="POR_htoct12"~"Morphological",
                             Traits=="BDX_htnov13"~"Morphological",
                             Traits=="AST_htnov12"~"Morphological",
                             Traits=="BDX_htnov18"~"Morphological",
                             Traits=="MeanBB"~"Phenological",
                             Traits=="MeanDBB"~"Phenological",
                             Traits=="POR_SLA"~"Functional",
                             Traits=="d13C"~"Functional"))


df.betaX$Traits <- as.factor(df.betaX$Traits)
levels(df.betaX$Traits)
df.betaX$Traits <- factor(df.betaX$Traits,levels=rev(c(traits)))
df.betaX$Xp <- as.factor(df.betaX$Xp)
df.betaX$Xp <- factor(df.betaX$Xp,levels=rev(levels(df.betaX$Xp)))

d13c <- parse(text = TeX("$\\delta^{13}C$"))

if(sub=="sub2"){
p <- df.betaX %>%   ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,show.legend = c(size = TRUE)) +
  facet_grid(.~Hyp,scales="free", space = "free") + 
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst")), # 
                     values=rev(c("#B2182B","#D73027", "#FDAE61","#5AAE61","#A6DBA0"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst")),
                     values = c(17, 17,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.27, .13),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank())

} else {
  
p <- df.betaX %>%   ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,show.legend = c(size = TRUE)) +
  facet_grid(.~Hyp,scales="free", space = "free") + 
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)), # 
                     values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)),
                     values = c(15, 15,17, 17,16,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.27, .13),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=2))}

p

if(sub=="sub1"){
  ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_Alldrivers_BetaX_PaperFig_sub1.png",height=8,width=12)} else if (sub=="sub2"){
  ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_Alldrivers_BetaX_PaperFig_sub2.png",height=8,width=10)} else {
  ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_Alldrivers_BetaX_PaperFig.png",height=8,width=12)  
  }
```

### Other visualizations I tried

```{r BetaXAllTraitsFigure}
sub="no" # this part was only one on the unfiltered dataset
traits <- c("POR_htoct12","AST_htnov12","BDX_htnov13","BDX_htnov18","MeanBB","MeanDBB","POR_SLA","d13C")
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT")    
param <- "betaX"

df.betaX <- mclapply(traits,ExtractParam) %>% bind_rows()

# Option 1
df.betaX$Traits <- as.factor(df.betaX$Traits)
df.betaX$Traits <- factor(df.betaX$Traits,levels=rev(c(traits)))
df.betaX$Xp <- as.factor(df.betaX$Xp)
df.betaX$Xp <- factor(df.betaX$Xp,levels=rev(levels(df.betaX$Xp)))


p <- df.betaX %>% 
  ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2) +
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  coord_flip() +
  scale_color_manual(values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9")),
                     labels = rev(c("Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Bordeaux, 85 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                parse(text = TeX("$\\delta^{13}C$")))
                     )) +
  theme_bw() +
  theme(axis.text = element_text(size=16),axis.title = element_text(size=16),
    legend.title=element_text(size=12), 
    legend.text=element_text(size=11),
    panel.grid.minor.y=element_blank(),
    panel.grid.major.y=element_blank()) +
    guides(color=guide_legend(reverse=TRUE))

p


# Option 2
p <- df.betaX %>% 
  mutate(Hyp=case_when(Xp=="A"~"Admixture",
                       Xp=="D"~"Admixture",
                       Xp=="EH1[20km]"~"Environmental heterogeneity",
                       Xp=="EH2[20km]"~"Environmental heterogeneity",
                       Xp=="EH1[1.6km]"~"Environmental heterogeneity",
                       Xp=="EH2[1.6km]"~"Environmental heterogeneity",
                       Xp=="SHM"~"Climate harshness",
                       Xp=="invEMT"~"Climate harshness")) %>% 
  ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Traits)) +
  facet_grid(rows = vars(Hyp),scales="free", space = "free") +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2) +
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  coord_flip() +
  scale_color_manual(values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9")),
                     labels = rev(c("Height (Portugal, 20 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Bordeaux, 85 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                parse(text = TeX("$\\delta^{13}C$")))
                     )) +
  theme_bw() +
  theme(axis.text = element_text(size=16),axis.title = element_text(size=16),
    legend.title=element_text(size=12), 
    legend.text=element_text(size=11),
    panel.grid.minor.y=element_blank(),
    panel.grid.major.y=element_blank()) +
    guides(color=guide_legend(reverse=TRUE))
p


df.betaX  <- df.betaX %>%
  mutate(Hyp=case_when(Xp=="A"~"Admixture",
                       Xp=="D"~"Admixture",
                       Xp=="EH1[20km]"~"Environmental heterogeneity",
                       Xp=="EH2[20km]"~"Environmental heterogeneity",
                       Xp=="EH1[1.6km]"~"Environmental heterogeneity",
                       Xp=="EH2[1.6km]"~"Environmental heterogeneity",
                       Xp=="SHM"~"Climate harshness",
                       Xp=="invEMT"~"Climate harshness"),
         TraitType=case_when(Traits=="POR_htoct12"~"Morphological",
                             Traits=="BDX_htnov13"~"Morphological",
                             Traits=="AST_htnov12"~"Morphological",
                             Traits=="BDX_htnov18"~"Morphological",
                             Traits=="MeanBB"~"Phenological",
                             Traits=="MeanDBB"~"Phenological",
                             Traits=="POR_SLA"~"Functional",
                             Traits=="d13C"~"Functional"))

# # Option 3
# 
# df.betaX$Xp <- factor(df.betaX$Xp,levels=rev(levels(df.betaX$Xp)))
# 
# pf.reverse <- df.betaX %>% 
#   ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Traits)) +
#   geom_pointinterval(position = position_dodge(width = .6),point_size=2.5) +
#   facet_grid(.~Hyp,scales="free", space = "free") + 
#   #facet_grid(~fct_relevel(Hyp,"Admixture","Climate harshness","Environmental heterogeneity"))
#   ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
#   scale_color_manual(values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9")),
#                      labels = rev(c("Height (Portugal, 20 months)",
#                                 "Height (Bordeaux, 25 months)",
#                                 "Height (Bordeaux, 85 months)",
#                                 "Height (Asturias, 21 months)",
#                                 "Specific Leaf Area",
#                                 "Mean bud burst date",
#                                 "Mean duration of bud burst",
#                                 parse(text = TeX("$\\delta^{13}C$")))
#                      )) +
#   theme_bw() +
#   theme(axis.text.x = element_text(size=10),
#         axis.text.y = element_text(size=16),
#         axis.title = element_text(size=16),
#         legend.title=element_text(size=12), 
#         strip.text.x = element_text(size = 14),
#         legend.text=element_text(size=9),
#         panel.grid.minor.x=element_blank(),
#         panel.grid.major.x=element_blank()) +
#     guides(color=guide_legend(reverse=TRUE))
#     
# ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_BetaX_FacetGrid_Reverse.png",height=8,width=14)

# Option 4
df.betaX$Xp <- factor(df.betaX$Xp,
                      levels=c("SHM","invEMT","EH1[20km]","EH1[1.6km]","EH2[20km]","EH2[1.6km]","D","A"))
df.betaX$Hyp <- as.factor(df.betaX$Hyp)
df.betaX$Hyp <- df.betaX$Hyp <- factor(df.betaX$Hyp,levels=c("Climate harshness","Admixture","Environmental heterogeneity"))


p <- df.betaX %>% 
  ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5) +
  facet_grid(.~Hyp,scales="free", space = "free") + 
  #facet_grid(~fct_relevel(Hyp,"Admixture","Climate harshness","Environmental heterogeneity"))
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  #coord_flip() +
  scale_color_manual(values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9")),
                     labels = rev(c("Height (Portugal, 20 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Bordeaux, 85 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                parse(text = TeX("$\\delta^{13}C$")))
                     )) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.112, .19),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(reverse=TRUE))
    
p


df.betaX$Hyp <- factor(df.betaX$Xp,
                      levels=c("SHM","invEMT","EH1[20km]","EH1[1.6km]","EH2[20km]","EH2[1.6km]","D","A"))


# Option 5
# For the paper, figure with not all the drivers
p <- df.betaX %>% 
  dplyr::filter(!(Xp %in% c("D","EH1[1.6km]","EH2[1.6km]"))) %>% 
  ggplot(aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5) +
  facet_grid(.~Hyp,scales="free", space = "free") + 
  #facet_grid(~fct_relevel(Hyp,"Admixture","Climate harshness","Environmental heterogeneity"))
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  #coord_flip() +
  scale_color_manual(values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9")),
                     labels = rev(c("Height (Portugal, 20 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Bordeaux, 85 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                parse(text = TeX("$\\delta^{13}C$")))
                     )) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.142, .185),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(reverse=TRUE))
    
p
```


## $\sigma^{2}_{C_{p}}$

```{r SigmaClon}
sub="no" # this part was only one on the unfiltered dataset
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT") 

provs <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
  dplyr::filter(!(prov=="MAD")) %>% 
  arrange(by=tree) %>% 
  distinct(prov) %>% 
  pull()

mclapply(traits,function(x){
if(x=="MeanBB"){
    legend.position <- c(0.85,0.86)} else {
    legend.position <- c(.15, .86)}
  
p <- readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("sigma2_clon"),
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, 
                rhat = F, 
                conf.int = T,
                conf.level = conf.level)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(Traits=x,
         prov=rep(provs,8)) %>% 
  ggplot(aes(x = prov, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=2,alpha=0.6,size=4) +
  scale_color_manual(values=c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("Populations") +
  ylab(TeX("$\\sigma^{2}_{C_{p}}$ estimates")) +
  labs(color = "Potential drivers") +
  theme(legend.position = legend.position,
        legend.background = element_rect(colour = "grey"),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=4))

ggsave(p,file=paste0("figs/PosteriorDistributions/",x,"_SigmaClon.png"),height=6,width=22)
  
})

```

## Broad-sense heritabilities

We calculated population-specific broad-sense heritabilities, such as:

$$ h^2_{p} = \frac{\sigma^{2}_{C_{p}}}{\sigma^{2}_{C_{p}} + \sigma^{2}_r} $$

```{r H2Figures}
sub="no" # this part was only one on the unfiltered dataset
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT") 

provs <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
  dplyr::filter(!(prov=="MAD")) %>% 
  arrange(by=tree) %>% 
  distinct(prov) %>% 
  pull()

mclapply(traits,function(x){
if(x=="MeanBB"){
    legend.position <- c(0.85,0.86)} else {
    legend.position <- c(.15, .86)}
  
p <- readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("h2_prov"),
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, 
                rhat = F, 
                conf.int = T,
                conf.level = conf.level)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(Traits=x,
         prov=rep(provs,8)) %>% 
  ggplot(aes(x = prov, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=2,alpha=0.6,size=4) +
  scale_color_manual(values=c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("Populations") +
  ylab(TeX("$\\H^{2}_{p}$ estimates")) +
  labs(color = "Potential drivers") +
  theme(legend.position = legend.position,
        legend.background = element_rect(colour = "grey"),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=4))

ggsave(p,file=paste0("figs/PosteriorDistributions/",x,"_h2prov.png"),height=6,width=22)
  
})

```

```{r H2pheno}
sub="no" # this part was only one on the unfiltered dataset

# For the discussion, I calculated the heritability values of the phenology-related traits:
h2pheno <- mclapply(c("MeanBB","MeanDBB"),function(x){
  tab <- readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("h2_prov"),
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, 
                rhat = F, 
                conf.int = T,
                conf.level = conf.level)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(Traits=x)
  
  mean(tab$estimate)
  })
```


## Variance partitioning


```{r VartPartFigs}
sub="no" # this part was only one on the unfiltered dataset

traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT") 

mclapply(traits,function(x){
  p <- readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("pi"),
                droppars = NULL, estimate.method = point.est, ess = F, rhat = F, conf.int = T,conf.level = 0.95)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(Traits=x,
         prop.var=rep(c("sigma_r","sigma_block","sigma_prov","mean_sigma_clon"),8)) %>% 
  
  ggplot(aes(x = prop.var, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=3,alpha=0.6,size=5) +
  scale_color_manual(values=c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("") +
  ylab("Proportion of variance explained") +
  labs(color = "Potential drivers") +
  scale_x_discrete(labels=c("Clones",
                            "Blocks",
                            "Populations",
                            "Residuals"
                            )) +
  theme(legend.position = c(.135, .81),
        legend.background = element_rect(colour = "grey"),
        axis.text = element_text(size=25),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=2))

ggsave(p,file=paste0("figs/PosteriorDistributions/",x,"_VarPart.png"),height=6,width=18)
  
})

```


```{r VarPartAllHeights}
sub="no" # this part was only one on the unfiltered dataset
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12")
drivers <- c("A","EH1[20km]","EH2[20km]","SHM","invEMT") 

sigmas <- mclapply(traits,function(x){
  readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("pi"),
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, rhat = F, 
                conf.int = T,
                conf.level = 0.95)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(Traits=x,
         prop.var=rep(c("sigma_r","sigma_block","sigma_prov","mean_sigma_clon"),8)) 
  
}) %>% 
  bind_rows() %>% 
  dplyr::filter(!(Drivers %in% c("D","EH.1km.PC1","EH.1km.PC2")))


maxylim <- 0.9
legend.text <- 20
line.size <- 10
point.size <- 5

p1 <- sigmas %>% 
  dplyr::filter(Traits=="POR_htoct12") %>% 
  ggplot(aes(x = prop.var, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=point.size,alpha=0.6,size=line.size) +
  scale_color_manual(values=c("#7FBC41",
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("") +
  ylim(0,maxylim) +
  ylab("Proportion of variance explained") +
  labs(color = "Potential drivers") +
  scale_x_discrete(labels=c("Clones",
                            "Blocks",
                            "Populations",
                            "Residuals"
                            )) +
  theme(legend.position = c(.2, .73),
        legend.text = element_text(size=legend.text,margin = margin(t = 10)),
        legend.background = element_rect(colour = "grey"),
        axis.text = element_text(size=22),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=1))

p2 <- sigmas %>% 
  dplyr::filter(Traits=="BDX_htnov13") %>% 
  ggplot(aes(x = prop.var, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=point.size,alpha=0.6,size=line.size) +
  scale_color_manual(values=c("#7FBC41",
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("") +
  ylab("") +
  ylim(0,maxylim) +
  labs(color = "Potential drivers") +
  scale_x_discrete(labels=c("Clones",
                            "Blocks",
                            "Populations",
                            "Residuals"
                            )) +
  theme(legend.position = "none",
        axis.text = element_text(size=22),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=1))


p3 <- sigmas %>% 
  dplyr::filter(Traits=="BDX_htnov18") %>% 
  ggplot(aes(x = prop.var, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=point.size,alpha=0.6,size=line.size) +
  scale_color_manual(values=c("#7FBC41",
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("") +
  ylim(0,maxylim) +
  ylab("Proportion of variance explained") +
  labs(color = "Potential drivers") +
  scale_x_discrete(labels=c("Clones",
                            "Blocks",
                            "Populations",
                            "Residuals"
                            )) +
  theme(legend.position = "none",
        axis.text = element_text(size=22),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=1))

p4 <- sigmas %>% 
  dplyr::filter(Traits=="AST_htnov12") %>% 
  ggplot(aes(x = prop.var, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=point.size,alpha=0.6,size=line.size) +
  scale_color_manual(values=c("#7FBC41",
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("") +
  ylab("") +
  ylim(0,maxylim) +
  labs(color = "Potential drivers") +
  scale_x_discrete(labels=c("Clones",
                            "Blocks",
                            "Populations",
                            "Residuals"
                            )) +
  theme(legend.position = "none",
        axis.text = element_text(size=22),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=1))

p <- plot_grid(p1,p2,p3,p4,
          labels = c('A)','B)',"C)","D)"),
                 label_size = 20 )

ggsave(p,file=paste0("figs/PosteriorDistributions/VarPart_AllHeights.png"),height=14,width=17)
  
```


## $\mathcal{R}^{2}$


```{r RsquaredPosteriors}
sub="no" # this part was only one on the unfiltered dataset
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT")    
param <- "R_squared"

df <- traits %>% 
  mclapply(ExtractParam) %>% 
  bind_rows() %>%
  mutate(trait.class=case_when(Traits=="POR_htoct12"~"Height",
                       Traits=="BDX_htnov13"~"Height",
                       Traits=="BDX_htnov18"~"Height",
                       Traits=="AST_htnov12"~"Height",
                       Traits=="MeanBB"~"Phenology",
                       Traits=="MeanDBB"~"Phenology",
                       Traits=="POR_SLA"~"Functional traits",
                       Traits=="d13C"~"Functional traits"),
         trait.names=case_when(Traits=="POR_htoct12"~"Portugal 20m",
                       Traits=="BDX_htnov13"~"Bordeaux 25m",
                       Traits=="BDX_htnov18"~"Bordeaux 85m",
                       Traits=="AST_htnov12"~"Asturias 21m",
                       Traits=="MeanBB"~"MeanBB",
                       Traits=="MeanDBB"~"MeanDBB",
                       Traits=="POR_SLA"~"SLA",
                       Traits=="d13C"~"d13C"))

df$Traits <- as.factor(df$Traits)
df$Traits <- factor(df$Traits,levels=c(traits))
df$Xp <- as.factor(df$Xp)
df$Xp <- factor(df$Xp,levels=rev(c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")))

p <- df %>% 
  ggplot(aes(x = trait.names, y = estimate,ymin = conf.low, ymax = conf.high,color=Xp)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,alpha=0.6) +
  facet_grid(.~trait.class,scales="free", space = "free") + 
  #facet_grid(~fct_relevel(Hyp,"Admixture","Climate harshness","Environmental heterogeneity"))
  ylab(TeX("$R^{2}$")) + xlab("") +
  labs(color = "Potential drivers") +
  scale_color_manual(values=rev(c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ))) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(.88, .19),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=2))
    
p

ggsave(p, file="figs/PosteriorDistributions/ALLTRAITS_Rsquared.png",height=7,width=12)
```



# Correlation btw the number of clones per pop and $\sigma^{2}_{C_{p}}$

```{r CorrealtionNumberClones}
sub="no" # this part was only one on the unfiltered dataset
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

provs <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
  dplyr::filter(!(prov=="MAD")) %>% 
  arrange(by=tree) %>% 
  distinct(prov) %>% 
  pull()

# Extrating the sigma2_clon (i.e. within-population genetic variation)
sigma2_clon <- mclapply(traits,function(x){
  
  readRDS(file=paste0("outputs/models/",x,".rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("sigma2_clon"),
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, 
                rhat = F, 
                conf.int = T,
                conf.level = conf.level)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(Traits=x,
         prov=rep(provs,8)) }) %>% 
    setNames(traits)


# Extracting the number of clones per population
NbClonePerProv <- function(trait){

  drivers <- readRDS(file="data/DF_Drivers.rds") %>% 
    dplyr::select(prov,A,D,EH.20km.PC1,EH.20km.PC2,EH.1km.PC1,EH.1km.PC2,SHM,invEMT) %>% 
    dplyr::mutate(across(-prov,scale))
  
  if(trait=="MeanBB"|trait=="MeanDBB"){
      
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    when(trait=="MeanBB" ~ dplyr::select(.,prov,clon,tree,block,BDX_s32013,BDX_s32014,BDX_s32015,BDX_s32017),
         ~ dplyr::select(.,prov,clon,tree,block,BDX_dbb2014,BDX_dbb2015,BDX_dbb2017)) %>% 
    filter(rowSums(is.na(across(where(is.numeric)))) != ncol(.)-4) %>% 
    dplyr::select(prov,clon) %>% 
    distinct(.) %>% 
    group_by(prov) %>% 
    dplyr::summarise(count=n())
  
  } else {
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    dplyr::rename(trait=all_of(trait)) %>% 
    dplyr::select(prov,clon,tree,block,trait) %>% 
    drop_na() %>%  
    dplyr::select(prov,clon) %>% 
    distinct(.) %>% 
    group_by(prov) %>% 
    dplyr::summarise(count=n())
  
  }}

count <- mclapply(traits,NbClonePerProv) %>% setNames(traits)

drivers <- c("A","D", "EH.20km.PC1","EH.20km.PC2","EH.1km.PC1","EH.1km.PC2","SHM","invEMT") 

df <- mclapply(traits, function(x){
  mclapply(drivers,function(driver)
  cor(count[[x]]$count,sigma2_clon[[x]]$estimate[sigma2_clon[[x]]$Drivers==driver])
  ) %>% 
  setNames(drivers)
}) %>% 
  setNames(traits) %>% 
  bind_rows(.id="Traits") %>% 
  gather(Driver, val, 2:ncol(.)) %>%
  spread(Traits, val) %>% 
  dplyr::select(Driver,POR_htoct12,BDX_htnov13,BDX_htnov18,AST_htnov12,MeanBB ,MeanDBB, POR_SLA, d13C) %>% 
  dplyr::mutate(Driver=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM"))


# Get the table in latex for the Supplementary Information
df %>% kable(digits=3,"latex")
```


# Climatic transfer distances

```{r FunctionDownloadDataCTD}
sub="no" # this part was only one on the unfiltered dataset

DownloadDataCTD <- function(trait){

if(trait=="MeanBB"|trait=="MeanDBB"|trait=="d13C"){
    CenterResponseVariable <- TRUE
    LogResponseVariable <- FALSE
    
  }  else if(trait=="POR_SLA"){
    CenterResponseVariable <- FALSE
    LogResponseVariable <- TRUE
  } else {
    CenterResponseVariable <- FALSE
    LogResponseVariable <- FALSE
  }
  
  drivers <- readRDS(file="data/DF_Drivers.rds") %>% 
    dplyr::select(prov,contains("_EMT"),contains("_SHM")) %>% 
    dplyr::mutate(dplyr::across(-prov,scale))
  
  if(trait=="MeanBB"|trait=="MeanDBB"){
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    when(trait=="MeanBB" ~ dplyr::select(.,prov,clon,tree,block,BDX_s32013,BDX_s32014,BDX_s32015,BDX_s32017),
         ~ dplyr::select(.,prov,clon,tree,block,BDX_dbb2014,BDX_dbb2015,BDX_dbb2017)) %>% 
    filter(rowSums(is.na(across(where(is.numeric)))) != ncol(.)-4) %>% 
    dplyr::mutate(trait = rowMeans(dplyr::select(.,contains("BDX")),na.rm=T)) %>%
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov") 
  
  } else {
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    dplyr::rename(trait=all_of(trait)) %>% 
    dplyr::select(prov,clon,tree,block,trait) %>% 
    drop_na() %>%  
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov")
  
  }
  
  # Depending on the site where the traits were measured, we do not use the same CTD
  if(trait %in% c("MeanBB","MeanDBB","BDX_htnov18","BDX_htnov13")) CTD <- c("CTD_bordeaux_EMT","CTD_bordeaux_SHM")
  if(trait %in% c("POR_SLA","d13C","POR_htoct12")) CTD <- c("CTD_portugal_EMT","CTD_portugal_SHM")
  if(trait == "AST_htnov12") CTD <- c("CTD_asturias_EMT","CTD_asturias_SHM")
  
  list.stan <- lapply(CTD, function(x)    {
    list(N=length(df$tree),
         y=df$trait,
         X=as.numeric(pull(unique(df[,c(x,"prov")]) %>% dplyr::select(contains(x)))),
         nprov = length(unique(df$prov)),
         nclon = length(unique(df$clon)),
         nblock = length(unique(df$block)),
         prov = as.numeric(as.factor(df$prov)),
         which_prov = as.numeric(as.factor(pull(unique(df[c("prov","clon")])[,"prov"]))),
         clon = as.numeric(as.factor(df$clon)),
         bloc = as.numeric(as.factor(df$block)))})
  
  names(list.stan) <- CTD
  
  return(list.stan)
}
```

```{r RunningModelsCTD}
sub="no" # this part was only one on the unfiltered dataset

traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

CTDbetaX <- lapply(traits, function(trait) {
  
  list.stan <- DownloadDataCTD(trait)
  
  
  output <- lapply(names(list.stan), function(x){
    
    mod <- sampling(model, 
                  data = list.stan[[x]], 
                  pars=c("betaX"),
                  iter = n_iter, 
                  chains = n_chains, 
                  cores = n_chains,
                  save_warmup = save_warmup,
                  thin=n_thin)
  
    broom::tidyMCMC(mod,pars=(c("betaX")),
                  droppars = NULL, estimate.method = "median", ess = T, rhat = T, conf.int = T,conf.level = 0.95) %>% 
      dplyr::mutate(driver=x)
    
  }) %>% bind_rows
  
  
  })

names(CTDbetaX) <- traits
CTDbetaX %>% bind_rows(.id="Trait")

saveRDS(CTDbetaX,file=paste0("outputs/models/CTDbetaX.rds"))
```


```{r FigureCTD, eval=T, fig.height=9,fig.width=9}
sub="no" # this part was only one on the unfiltered dataset

CTDbetaX <- readRDS(file=paste0("outputs/models/CTDbetaX.rds"))

traits <- c("BDX_htnov18","BDX_htnov13","POR_htoct12","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

CTDbetaX <- 
  CTDbetaX %>% 
  bind_rows(.id="Traits") %>%
  mutate(driver=paste0("CMT(",str_sub(driver,-3),")"))

CTDbetaX $Traits <- as.factor(CTDbetaX $Traits)
CTDbetaX $Traits <- factor(CTDbetaX $Traits,levels=rev(c(traits)))

d13c <- parse(text = TeX("$\\delta^{13}C$"))

p <- CTDbetaX  %>%   ggplot(aes(x = driver, y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,size=3,show.legend = c(size = TRUE)) +
  #facet_grid(.~Hyp,scales="free", space = "free") + 
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)), # 
                     values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)),
                     values = c(15, 15,17, 17,16,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.position = c(0.83,0.84),#c(.37, .16),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=1,override.aes = list(size=2)))
p

ggsave(p,file="figs/PosteriorDistributions/CTD.png",height=9,width=9)
```


# Genetic diversity

We want to estimate the relationship between within-population quantitative genetic variation and genetic diversity (*expected Nei’s heterozygosity* $H_e$), to check that demographic factors do not influence quantitative genetic variation. For that, we take the Nei’s heterozygosity $H_e$ values from @rodriguez2016capturing, for both nuSSRs (12 nuclear microsatellites) and SNPs (266 SNP markers).

```{r GeneticDiversityRodriguezQuilon}
drivers <- readRDS(file="data/DF_Drivers.rds") %>% 
    dplyr::select(prov) %>% 
    dplyr::mutate(He.nuSSRs=case_when(prov=="HOU" ~ 0.55,
                                      prov=="VER" ~ 0.57,
                                      prov=="MIM" ~ 0.58,
                                      prov=="OLO" ~ 0.57,
                                      prov=="PET" ~ 0.55,
                                      prov=="PLE" ~ 0.61,
                                      prov=="STJ" ~ 0.60,
                                      prov=="ALT" ~ 0.66,
                                      prov=="ARM" ~ 0.56,
                                      prov=="CAD" ~ 0.55,
                                      prov=="CAS" ~ 0.54,
                                      prov=="LAM" ~ 0.50,
                                      prov=="LEI" ~ 0.62,
                                      prov=="PUE" ~ 0.54,
                                      prov=="SAC" ~ 0.59,
                                      prov=="SEG" ~ 0.56,
                                      prov=="SIE" ~ 0.56,
                                      prov=="ARN" ~ 0.58,
                                      prov=="BAY" ~ 0.58,
                                      prov=="BON" ~ 0.57,
                                      prov=="CAR" ~ 0.65,
                                      prov=="CEN" ~ 0.56,
                                      prov=="COC" ~ 0.60,
                                      prov=="CUE" ~ 0.63,
                                      prov=="OLB" ~ 0.58,
                                      prov=="QUA" ~ 0.64,
                                      prov=="SAL" ~ 0.58,
                                      prov=="VAL" ~ 0.55,
                                      prov=="PIA" ~ 0.52,
                                      prov=="PIE" ~ 0.52,
                                      prov=="COM" ~ 0.76,
                                      prov=="ORI" ~ 0.64,
                                      prov=="TAM" ~ 0.55),
                    He.SNPs=case_when(prov=="HOU" ~ 0.25,
                                      prov=="VER" ~ 0.26,
                                      prov=="MIM" ~ 0.26,
                                      prov=="OLO" ~ 0.25,
                                      prov=="PET" ~ 0.26,
                                      prov=="PLE" ~ 0.27,
                                      prov=="STJ" ~ 0.26,
                                      prov=="ALT" ~ 0.24,
                                      prov=="ARM" ~ 0.23,
                                      prov=="CAD" ~ 0.24,
                                      prov=="CAS" ~ 0.24,
                                      prov=="LAM" ~ 0.23,
                                      prov=="LEI" ~ 0.26,
                                      prov=="PUE" ~ 0.23,
                                      prov=="SAC" ~ 0.24,
                                      prov=="SEG" ~ 0.25,
                                      prov=="SIE" ~ 0.26,
                                      prov=="ARN" ~ 0.29,
                                      prov=="BAY" ~ 0.28,
                                      prov=="BON" ~ 0.30,
                                      prov=="CAR" ~ 0.27,
                                      prov=="CEN" ~ 0.27,
                                      prov=="COC" ~ 0.29,
                                      prov=="CUE" ~ 0.28,
                                      prov=="OLB" ~ 0.29,
                                      prov=="QUA" ~ 0.29,
                                      prov=="SAL" ~ 0.28,
                                      prov=="VAL" ~ 0.28,
                                      prov=="PIA" ~ 0.25,
                                      prov=="PIE" ~ 0.25,
                                      prov=="COM" ~ 0.30,
                                      prov=="ORI" ~ 0.29,
                                      prov=="TAM" ~ 0.19)) %>% 
    dplyr::mutate(dplyr::across(-prov,scale))
```

```{r FunctionFormatGeneticDiversity}
FormatGeneticDiversity <- function(trait){

if(trait=="MeanBB"|trait=="MeanDBB"|trait=="d13C"){
    CenterResponseVariable <- TRUE
    LogResponseVariable <- FALSE
    
  }  else if(trait=="POR_SLA"){
    CenterResponseVariable <- FALSE
    LogResponseVariable <- TRUE
  } else {
    CenterResponseVariable <- FALSE
    LogResponseVariable <- FALSE
  }
  
  if(trait=="MeanBB"|trait=="MeanDBB"){
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    when(trait=="MeanBB" ~ dplyr::select(.,prov,clon,tree,block,BDX_s32013,BDX_s32014,BDX_s32015,BDX_s32017),
         ~ dplyr::select(.,prov,clon,tree,block,BDX_dbb2014,BDX_dbb2015,BDX_dbb2017)) %>% 
    filter(rowSums(is.na(across(where(is.numeric)))) != ncol(.)-4) %>% 
    dplyr::mutate(trait = rowMeans(dplyr::select(.,contains("BDX")),na.rm=T)) %>%
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov") 
  
  } else {
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    dplyr::rename(trait=all_of(trait)) %>% 
    dplyr::select(prov,clon,tree,block,trait) %>% 
    drop_na() %>%  
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov")
  
  }
  
  var.He <- c("He.nuSSRs","He.SNPs")
  
  list.stan <- lapply(var.He, function(x)    {
    list(N=length(df$tree),
         y=df$trait,
         X=as.numeric(pull(unique(df[,c(x,"prov")]) %>% dplyr::select(contains(x)))),
         nprov = length(unique(df$prov)),
         nclon = length(unique(df$clon)),
         nblock = length(unique(df$block)),
         prov = as.numeric(as.factor(df$prov)),
         which_prov = as.numeric(as.factor(pull(unique(df[c("prov","clon")])[,"prov"]))),
         clon = as.numeric(as.factor(df$clon)),
         bloc = as.numeric(as.factor(df$block)))})
  
  names(list.stan) <- var.He
  
  return(list.stan)
}
```



```{r RunningModelsGeneticDiversity}
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

GenDivbetaX <- lapply(traits, function(trait) {
  
  list.stan <- FormatGeneticDiversity(trait)
  
  
  output <- lapply(names(list.stan), function(x){
    
    mod <- sampling(model, 
                  data = list.stan[[x]], 
                  pars=c("betaX"),
                  iter = n_iter, 
                  chains = n_chains, 
                  cores = n_chains,
                  save_warmup = save_warmup,
                  thin=n_thin)
  
    broom::tidyMCMC(mod,pars=(c("betaX")),
                  droppars = NULL, estimate.method = "median", ess = T, rhat = T, conf.int = T,conf.level = 0.95) %>% 
      dplyr::mutate(driver=x)
    
  }) %>% bind_rows
  
  
  })

names(GenDivbetaX) <- traits
GenDivbetaX %>% bind_rows(.id="Trait")

saveRDS(GenDivbetaX,file=paste0("outputs/models/GenDivbetaX.rds"))
```


```{r FigureGeneticDiversity, eval=T, fig.height=9,fig.width=13}
GenDivbetaX <- readRDS(file=paste0("outputs/models/GenDivbetaX.rds"))

traits <- c("BDX_htnov18","BDX_htnov13","POR_htoct12","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

GenDivbetaX <- 
  GenDivbetaX %>% 
  bind_rows(.id="Traits")

GenDivbetaX$Traits <- as.factor(GenDivbetaX$Traits)
GenDivbetaX$Traits <- factor(GenDivbetaX$Traits,levels=rev(c(traits)))

d13c <- parse(text = TeX("$\\delta^{13}C$"))

p <- GenDivbetaX  %>%   ggplot(aes(x = driver, y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,size=3,show.legend = c(size = TRUE)) +
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)), # 
                     values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)),
                     values = c(15, 15,17, 17,16,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      legend.background = element_rect(colour = "grey"),
      strip.text.x = element_text(size = 14),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=1,override.aes = list(size=2)))
p

ggsave(p,file="figs/PosteriorDistributions/GeneticDiversity_betaX.png",height=9,width=13)
```

# Isolation-by-distance

Following the comments of referee #1 (first round of review in Heredity), we calculate an index to quantify (or at least try to capture) the geographic isolation of the populations within the species range. 
For that, we use the distribution of maritime pine based on both the Euforgen distribution and the NFI plots of Spain and France. We define the center of the distribution as the point at mid-distance of the most extreme latitudes and longitudes where the presence of maritime pine has been observed. 
Here are some maps showing maritime pine distribution, the population location and the center of the distribution.

```{r MappingCenterDistributionIBD}
library(raster)
library(rnaturalearth)
library(ggthemes)
library(sf)


# We extract the population coordinates
prov.coord <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
  dplyr::select(prov, longitude_prov,latitude_prov) %>% 
  distinct() %>% 
  dplyr::filter(!prov=="MAD") %>% 
  dplyr::rename(longitude=longitude_prov,latitude=latitude_prov)


PinpinDistri  <- shapefile('../../GenomicOffset/GenomicOffsetPinPin/data/maps/MaskPinpinDistri/PinpinDistriEUforgen_NFIplotsBuffer10km.shp')
center.distri.long <- mean(c(extent(PinpinDistri)@xmin,extent(PinpinDistri)@xmax))
center.distri.lat <- mean(c(extent(PinpinDistri)@ymin,extent(PinpinDistri)@ymax))
center.distri <- c(center.distri.long,center.distri.lat)


# Maps with the raster package
##############################

# Create spatial objects for plotting
prov.coord.spatial <- SpatialPoints(prov.coord[,c("longitude","latitude")], 
                        proj4string=CRS("+proj=longlat +datum=WGS84"))
center.distri.spatial <- SpatialPoints(data.frame(longitude=center.distri.long,latitude=center.distri.lat), 
                        proj4string=CRS("+proj=longlat +datum=WGS84"))

plot(PinpinDistri,col="cadetblue1", lwd=0.1)
plot(prov.coord.spatial,add=T,pch=17,col="darkblue")
plot(center.distri.spatial,add=T,pch=4,col="red",lwd=4)


# Maps with the sf and ggplot packages
######################################

# This map is included in the Supplementary Information and the response to the reviewers.

world <- ne_countries(scale = "medium", returnclass = "sf")

PinpinDistri <- st_read('../../GenomicOffset/GenomicOffsetPinPin/data/maps/MaskPinpinDistri/PinpinDistriEUforgen_NFIplotsBuffer10km.shp')
mapIBD <- ggplot(PinpinDistri) +
  geom_sf(color=alpha("chartreuse4",0.6),size=0.2,fill=alpha("chartreuse4",0.3)) + 
  theme_bw() +
  geom_sf(data = world,color=alpha("gray",0.8),size=0.2,fill=alpha("gray",0.1)) +
    coord_sf(xlim = c(-10, 13), 
             ylim = c(32,50.5),
             crs = "+proj=longlat +datum=WGS84") +
  geom_point(data=prov.coord[,c("longitude","latitude")],aes(x=longitude,y=latitude),size=2,pch=17) +
  geom_point(data=data.frame(longitude=center.distri.long,latitude=center.distri.lat),aes(x=longitude,y=latitude),size=6,pch=8,col="red") +
  xlab("Longitude") + ylab("Latitude")

ggsave(mapIBD,file="maps/mapIBD.png",height=8,width=10)
```

We then calculate the distance (in meters) between each population location and the center of the distribution. With the `geosphere` package.

```{r FormatDataIBD}
# calculate the distance to the point with the package geosphere
library(geosphere)
drivers <- prov.coord %>% 
  dplyr::mutate(dist.center.pt = distGeo(center.distri, prov.coord[,c("longitude","latitude")])) %>% 
  dplyr::mutate(dist.center.pt.scaled = as.numeric(scale(dist.center.pt)))

FormatIBD <- function(trait){

if(trait=="MeanBB"|trait=="MeanDBB"|trait=="d13C"){
    CenterResponseVariable <- TRUE
    LogResponseVariable <- FALSE
    
  }  else if(trait=="POR_SLA"){
    CenterResponseVariable <- FALSE
    LogResponseVariable <- TRUE
  } else {
    CenterResponseVariable <- FALSE
    LogResponseVariable <- FALSE
  }
  
  if(trait=="MeanBB"|trait=="MeanDBB"){
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    when(trait=="MeanBB" ~ dplyr::select(.,prov,clon,tree,block,BDX_s32013,BDX_s32014,BDX_s32015,BDX_s32017),
         ~ dplyr::select(.,prov,clon,tree,block,BDX_dbb2014,BDX_dbb2015,BDX_dbb2017)) %>% 
    filter(rowSums(is.na(across(where(is.numeric)))) != ncol(.)-4) %>% 
    dplyr::mutate(trait = rowMeans(dplyr::select(.,contains("BDX")),na.rm=T)) %>%
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov") 
  
  } else {
    
  df <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    dplyr::rename(trait=all_of(trait)) %>% 
    dplyr::select(prov,clon,tree,block,trait) %>% 
    drop_na() %>%  
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov")
  
  }
  
  
  list.stan <- list(N=length(df$tree),
         y=df$trait,
         X=unique(df$dist.center.pt.scaled),
         nprov = length(unique(df$prov)),
         nclon = length(unique(df$clon)),
         nblock = length(unique(df$block)),
         prov = as.numeric(as.factor(df$prov)),
         which_prov = as.numeric(as.factor(pull(unique(df[c("prov","clon")])[,"prov"]))),
         clon = as.numeric(as.factor(df$clon)),
         bloc = as.numeric(as.factor(df$block)))
  
  return(list.stan)
}
```

We run one model per trait (so eight models) and we extract the $\beta_X$ coefficients standing for the association between the total genetic variances and the distance between each population and the center of the species distribution.

```{r RunModelsIBD}
traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

IBDbetaX <- lapply(traits, function(trait) {
  
  list.stan <- FormatIBD(trait)
  
  mod <- sampling(model, 
                  data = list.stan, 
                  pars=c("betaX"),
                  iter = n_iter, 
                  chains = n_chains, 
                  cores = n_chains,
                  save_warmup = save_warmup,
                  thin=n_thin)
  
  broom::tidyMCMC(mod,pars=(c("betaX")),
                  droppars = NULL, estimate.method = "median", ess = T, rhat = T, conf.int = T,conf.level = 0.95) 
    
  }) %>% 
  bind_rows() %>% 
  dplyr::mutate(Traits=traits)
  

saveRDS(IBDbetaX,file=paste0("outputs/models/IBDbetaX.rds"))
```

We build the figure included in the Supplementary Information with the $\beta_X$ coefficients.

```{r BetaXIBD, eval=T, fig.height=9,fig.width=13}
IBDbetaX <- readRDS(file=paste0("outputs/models/IBDbetaX.rds"))

traits <- c("BDX_htnov18","BDX_htnov13","POR_htoct12","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")


IBDbetaX$Traits <- as.factor(IBDbetaX$Traits)
IBDbetaX$Traits <- factor(IBDbetaX$Traits,levels=rev(c(traits)))

d13c <- parse(text = TeX("$\\delta^{13}C$"))

p <- IBDbetaX  %>%   ggplot(aes(x=term,y = estimate,ymin = conf.low, ymax = conf.high,colour=Traits,shape=Traits)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=2.5,size=3,show.legend = c(size = TRUE)) +
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_colour_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)), # 
                     values=rev(c("#B2182B","#D73027","#F46D43", "#FDAE61","#5AAE61","#A6DBA0" ,"#4575B4", "#ABD9E9"))) +
  scale_shape_manual(labels = rev(c("Height (Bordeaux, 85 months)",
                                "Height (Bordeaux, 25 months)",
                                "Height (Portugal, 20 months)",
                                "Height (Asturias, 21 months)",
                                "Mean bud burst date",
                                "Mean duration of bud burst",
                                "Specific Leaf Area",
                                d13c)),
                     values = c(15, 15,17, 17,16,16,16,16)) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=16),
        axis.title = element_text(size=16),
      legend.title=element_text(size=13), 
      legend.text=element_text(size=10),
      axis.ticks.x = element_blank(),
      legend.background = element_rect(colour = "grey"),
      legend.position = c(.85, .15),
      strip.text.x = element_blank(),
      panel.grid.minor.x=element_blank(),
      panel.grid.major.x=element_blank()) +
    guides(color=guide_legend(ncol=1,override.aes = list(size=2)))

p

ggsave(p,file="figs/PosteriorDistributions/IBD_betaX.png",height=9,width=10)
```

# Experimental design 

```{r ExpDesign}
sub="no" # this part was only one on the unfiltered dataset

traits <- c("POR_htoct12","BDX_htnov13","BDX_htnov18","AST_htnov12","MeanBB","MeanDBB","POR_SLA","d13C")

list.df.traits <- lapply(traits, function(x){
  list.trait <- DownloadData(x,sub=sub)
  return(list.trait$df %>% mutate(trait.id=x))})


lapply(list.df.traits, function(x){
  
# Minimum, maximum and average number of trees per clone in each population
df <- x %>% 
    dplyr::select(prov,clon) %>% 
    group_by(clon) %>% 
    mutate(NbInd = n()) %>% 
    distinct() %>% 
    group_by(prov) %>% 
    mutate(MinNbTreesPerClone=min(NbInd),MaxNbTreesPerClone=max(NbInd),MeanNbTreesPerClone=mean(NbInd)) %>% 
    dplyr::select(-clon,-NbInd) %>% 
    distinct() 

# Number of trees in each population
df <- x %>% 
    group_by(prov) %>% 
    summarise(NbTrees=n()) %>% 
    merge(df,by="prov")
  
# Number of clones in each population  
df <- x %>% 
    dplyr::select(prov,clon) %>% 
    distinct() %>% 
    group_by(prov) %>%  
    summarise(NbClones=n()) %>% 
    merge(df,by="prov")
  
# Export in a latex table for the Supplementary Information
df <- df %>% dplyr::rename(Population=prov) 

# print(xtable(df, type = "latex",digits=1), 
#       file = paste0("tables/ExperimentalDesign/",unique(x$trait.id),".tex"),
#       include.rownames=FALSE)
  })
```

# Model with G*E

## Model equation and code

This model is based on the previous one (so I do not re-explain the meaning of each parameters, except for the new ones). 

We model each trait $y_{bpcrs}$ such as:

\begin{equation}
\begin{aligned}
y_{bpcr} & \sim  \mathcal{N}(\mu_{bpc},\sigma^{2}_{r})\\[3pt]
\mu_{bpc} & =  \beta_{0} +  B_{b} + P_{p} + C_{c(p)} + S_{s} +  S_{s}*C_{c(p)}\\[3pt]
\beta_{0} & \sim \mathcal{N}(\mu_{y},2)\\[3pt]
\begin{bmatrix} B_{b} \\ P_{p} \\ S_{s} \\ S_{s}*C_{c(p)} \end{bmatrix} 
    & \sim \mathcal{N}\left(0,
\begin{bmatrix}\sigma^{2}_{B}\\[3pt]
\sigma^{2}_{P}\\[3pt]
\sigma^{2}_{S}\\[3pt]
\sigma^{2}_{S*C}\\[3pt]
\end{bmatrix}
\right)\\[3pt]
C_{c(p)} & \sim \mathcal{N}(0,\sigma^{2}_{C_{p}})
\end{aligned}
\end{equation}

where $S_{s}$ are the site intercepts, $\sigma^{2}_{S}$ is the variance among sites and $\sigma^{2}_{S*C}$ is the variance of the interaction among the sites and clones, that is the G*E variance.

Partitionning of the total variance $\sigma_{tot}^{2}$:

\begin{equation}
\begin{aligned}
\sigma_{tot}^{2} & = \sigma_{r}^{2} + \sigma_{B}^{2} + \overline{\sigma_{C_{p}}^{2}} + \sigma_{P}^{2} + \sigma_{S}^{2} + \sigma_{S*C}^{2}\\[3pt]
\sigma_{r} & = \sigma_{tot} \times \sqrt(\pi_{r})\\[3pt]
\sigma_{B} & = \sigma_{tot} \times \sqrt(\pi_{B})\\[3pt]
\sigma_{P} & = \sigma_{tot} \times \sqrt(\pi_{P})\\[3pt]
\overline{\sigma_{C_{p}}} & = \sigma_{tot} \times \sqrt(\pi_{C})\\[3pt]
\sigma_{S} & = \sigma_{tot} \times \sqrt(\pi_{S})\\[3pt]
\sigma_{S*C} & = \sigma_{tot} \times \sqrt(\pi_{S*C})\\[3pt]
\sigma_{tot} & \sim \mathcal{S}^{*}(0,1,3)
\end{aligned}
\end{equation}

with $\sum_{l}^{6}\pi_{l}=1$.

The expression of $\sigma_{C_{p}}$ (and its associated priors) is the same than the main model of the paper (which is trait-specific).

Here is the stan model code:

```{r ModelGECompilation}
model = stan_model("scripts/StanModels/HierarchicalModel_VaryingInterClonesSD_OneCovariate_3Sites.stan")
model
```

## Running the GE model

```{r FormatDataForGEModel}
# The three height measurements we used:
traits <- c("POR_htoct12","BDX_htnov13","AST_htnov12")

# Height is not transformed
CenterResponseVariable <- FALSE
LogResponseVariable <- FALSE


# Downloading and scaling (mean-centering) the potential drivers
drivers <- readRDS(file="data/DF_Drivers.rds") %>% 
    dplyr::select(prov,A,D,EH.20km.PC1,EH.20km.PC2,EH.1km.PC1,EH.1km.PC2,SHM,invEMT) %>% 
    dplyr::mutate(across(-prov,scale)) %>% 
    dplyr::mutate(across(-prov,as.numeric))

  
# Merging with the phenotypic data
df <- lapply(traits, function(trait){
   readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
    dplyr::filter(!(prov=="MAD")) %>% 
    dplyr::rename(trait=all_of(trait)) %>% 
    dplyr::select(prov,clon,tree,block,site,trait) %>% 
    dplyr::mutate(site_clon=paste0(site,"_",clon)) %>% 
    drop_na() %>%  
    execute_if(CenterResponseVariable, mutate(trait=trait-mean(trait))) %>% 
    execute_if(LogResponseVariable, mutate(trait=log(trait))) %>% 
    arrange(by=tree) %>% 
    inner_join(drivers,by="prov")
}) %>% bind_rows()


list.stan <- lapply(colnames(drivers %>% dplyr::select(-prov)), function(x) {
    list(N=length(df$tree),
         y=df$trait,
         X=as.numeric(pull(unique(df[,c(x,"prov")]) %>% dplyr::select(contains(x)))),
         nprov = length(unique(df$prov)),
         nclon = length(unique(df$clon)),
         nblock = length(unique(df$block)),
         nsite = length(unique(df$site)),
         prov = as.numeric(as.factor(df$prov)),
         which_prov = as.numeric(as.factor(pull(unique(df[c("prov","clon")])[,"prov"]))),
         clon = as.numeric(as.factor(df$clon)),
         bloc = as.numeric(as.factor(df$block)),
         site = as.numeric(as.factor(df$site)),
         site_clon = as.numeric(as.factor(df$site_clon)))})
  
names(list.stan) <- colnames(drivers %>% dplyr::select(-prov))
```


```{r RunGEModel}
list.models <- list()

for(i in names(list.stan)){
  
list.models[[i]] <- sampling(model, 
                               data = list.stan[[i]], 
                               pars=c("betaX", "pi", "R_squared", "h2_prov"), # we keep only the parameters of interest to save space
                               iter = n_iter, 
                               chains = n_chains, 
                               cores = n_chains,
                               save_warmup = save_warmup,
                               thin=n_thin)}

saveRDS(list.models,file=paste0("outputs/models/HeightGE.rds"))
```


## Figures of post. distributions

### $\beta_{X}$
```{r BetaXGEModel, eval=T,fig.height=8,fig.width=12}
drivers <- c("A","D","EH1[20km]","EH2[20km]","EH1[1.6km]","EH2[1.6km]","SHM","invEMT")# order in the model list, do not change

d <- readRDS(file=paste0("outputs/models/HeightGE.rds")) %>% 
  mclapply(function(x) broom::tidyMCMC(x,pars=("betaX"),
              droppars = NULL, 
              estimate.method = point.est, 
              ess = F, 
              rhat = F, 
              conf.int = T,
              conf.level = conf.level)) %>% 
  bind_rows() %>% 
  mutate(Xp=drivers)
  
d$Xp <- as.factor(d$Xp)
d$Xp <- factor(d$Xp,levels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM"))


p <- ggplot(d,aes(x = Xp, y = estimate,ymin = conf.low, ymax = conf.high,color=Xp)) +
  geom_pointinterval(position = position_dodge(width = .6),point_size=5,size=5) +
  ylab(TeX("$\\beta_{X}$ estimates")) + xlab("") +
  scale_x_discrete(labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +

  scale_color_manual(values=c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" )) +
  theme_bw() +
  theme(axis.text = element_text(size=16),
        axis.title = element_text(size=16),
        legend.position = "none")

ggsave(p,file=paste0("figs/PosteriorDistributions/HeightGE_BetaX.png"),height=8,width=12)
p
```

### Broad-sense heritabilities

```{r H2GEModel, eval=T,fig.height=6,fig.width=22}
provs <- readRDS(file="data/ClonapinData/PhenoDataNovember2019_AnnualTraits.rds") %>% 
  dplyr::filter(!(prov=="MAD")) %>% 
  arrange(by=tree) %>% 
  distinct(prov) %>% 
  pull()

legend.position <- c(.15, .86)
  
p <- readRDS(file=paste0("outputs/models/HeightGE.rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("h2_prov"),
                droppars = NULL, 
                estimate.method = point.est, 
                ess = F, 
                rhat = F, 
                conf.int = T,
                conf.level = conf.level)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(prov=rep(provs,8)) %>% 
  ggplot(aes(x = prov, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=2,alpha=0.6,size=4) +
  scale_color_manual(values=c("#7FBC41", "#4D9221" ,
                              "#DE77AE","#C51B7D" ,
                              "#B2ABD2", "#542788" , #"#2D004B","#8073AC" ,
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("Populations") +
  ylab(TeX("$\\H^{2}_{p}$ estimates")) +
  labs(color = "Potential drivers") +
  theme(legend.position = legend.position,
        legend.background = element_rect(colour = "grey"),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=4))

ggsave(p,file=paste0("figs/PosteriorDistributions/HeightGE_h2prov.png"),height=6,width=22)
p
```


### Variance partitioning

```{r VartPartGEModel, eval=T, fig.height=6,fig.width=18}
p <- readRDS(file=paste0("outputs/models/HeightGE.rds")) %>% 
  mclapply(function(x) {
    broom::tidyMCMC(x,pars=("pi"),
                droppars = NULL, estimate.method = point.est, ess = F, rhat = F, conf.int = T,conf.level = 0.95)}) %>%  
  bind_rows(.id="Drivers") %>% 
  mutate(prop.var=rep(c("sigma_r","sigma_block","sigma_prov","mean_sigma_clon","sigma_site","sigma_GE"),8),
         prop.var=factor(prop.var, levels=c("sigma_prov","mean_sigma_clon","sigma_GE","sigma_site","sigma_block","sigma_r"))) %>%
  ggplot(aes(x = prop.var, y = estimate,ymin = conf.low, ymax = conf.high,color=Drivers)) +
  geom_pointinterval(position = position_dodge(width = .8),point_size=3,alpha=0.6,size=5) +
  scale_color_manual(values=c("#7FBC41", "#4D9221",
                              "#DE77AE","#C51B7D",
                              "#B2ABD2", "#542788",
                              "#E08214" ,"#FDB863" ),
                     labels=c("A","D","EH1[1.6km]","EH2[1.6km]","EH1[20km]","EH2[20km]","invEMT","SHM")) +
  xlab("") +
  ylab("Proportion of variance explained") +
  labs(color = "Potential drivers") +
  scale_x_discrete(labels=c("Populations",
                            "Clones",
                            "Sites*Clones",
                            "Sites",
                            "Blocks",
                            "Residuals"
                            )) +
  theme(legend.position = c(.135, .81),
        legend.background = element_rect(colour = "grey"),
        axis.text = element_text(size=25),
        panel.grid.minor.x=element_blank(),
        panel.grid.major.x=element_blank())  +
    guides(color=guide_legend(ncol=2))

ggsave(p,file=paste0("figs/PosteriorDistributions/HeightGE_VarPart.png"),height=6,width=18)
p
```



# Bibliography


